torchvision/datasets/stl10.py

```diff
@@ -182,4 +182,6 @@ def __load_folds(self, folds: Optional[int]) -> None:
         with open(path_to_folds, 'r') as f:
             str_idx = f.read().splitlines()[folds]
             list_idx = np.fromstring(str_idx, dtype=np.uint8, sep=' ')
-            self.data, self.labels = self.data[list_idx, :, :, :], self.labels[list_idx]
+            self.data = self.data[list_idx, :, :, :]
+            if self.labels is not None:
```

Comment:
Can `self.labels` actually be None in this code path?

If it can, then mypy properly caught a potential bug and we should probably add a test for it.

If it can't, `mypy` is unfortunately forcing us to obfuscate the code (as it often does...)

Reply:
Yes it can be None. See:
https://github.com/pytorch/vision/blob/5196845e1194a8c3170da567e29cc58bcc4c49ff/torchvision/datasets/stl10.py#L140-L141

@pmeier: Why are there no tests for the entire STL10 class? I think this is something that could be addressed on a separate PR, this one focuses on fixing master failures.

Reply:
@datumbox There are no tests for a lot of datasets, see https://github.com/pytorch/vision/issues/963#issuecomment-770914668. I'm on it.

Reply:
The type annotation of the function doesn't tell us much unfortunately. It tells us that `self.labels` can be None in general, but it doesn't tell whether it can be `None` in that particular code path.

As far as I can tell, `__load_folds` is only called right after `__loadfile` is called with  `labels_file`, and so `self.labels` can never be `None` within `__load_folds`.
 Unfortunately, this isn't something that `mypy` can figure out, but I feel like the hurts readability of the code.

Perhaps we should tell mypy to ignore this line instead?

Reply:
I agree that based on the current code the `self.labels` seems to always has values. Other parts of the class seem to also check for on whether `self.labels` is None, so I think it's worth maintaining it:
https://github.com/pytorch/vision/blob/5196845e1194a8c3170da567e29cc58bcc4c49ff/torchvision/datasets/stl10.py#L120-L123

It might be worth refactoring this on a separate PR to simplify the logic and increase the readability.

Reply:
> Other parts of the class seem to also check for on whether self.labels is None

It makes sense to check in these parts because `self.labels` can indeed be None there (although I haven't double-checked).

But checking for `None` in `__load_folds` is misleading IMHO:

- it suggests that `labels` can be None even though it can't
- it doesn't say what to do if `labels` is None, i.e. there's no `else` clause.

Reply:
@NicolasHug Sounds good. The scope of this PR is not to do massive rewrites on the dataset class bur rather fix the failing master and facilitate FBcode syncing. Please create an issue with your proposals describing which parts you think should be rewritten/simplified so that we can discuss it in more detail.

Reply:
I'm not suggesting to refactor the code, I'm suggesting to ignore mypy